training/src/main/scala/com/airbnb/aerosolve/training/pipeline/NDTreePipeline.scala

```diff
@@ -0,0 +1,194 @@
+        val stats = feature.asInstanceOf[FeatureStats]
+        log.info(s"${family}, ${name}: ${stats.count} ${stats.min} ${stats.max}")
+      } else {
+        val model = feature.asInstanceOf[NDTreeModel]
+        val nodes = model.getNodes
+        log.info(s"${family}, ${name}: ${model.getDimension} ${nodes.length} ${nodes.mkString("\n")}")
+      }
+    }
+  }
+
+  def buildFeatures(sc: SparkContext, config : Config):
+      Array[((String, String), Any)] = {
+    val cfg = config.getConfig("make_feature_map")
+    val inputPattern: String = cfg.getString("input")
+    log.info(s"Training data: ${inputPattern}")
+    val sample: Double = cfg.getDouble("sample")
+    val minCount: Int = cfg.getInt("min_count")
+    // minMax.filter(x => !linearFeatureFamilies.contains(x._1._1))
+    val linearFeatureFamilies: java.util.List[String] = Try(config.getStringList("linear_feature"))
+      .getOrElse[java.util.List[String]](List.empty.asJava)
+
+    val input = GenericPipeline.getExamples(sc, inputPattern)
+    val options = NDTreeBuildOptions(
+      maxTreeDepth = cfg.getInt("max_tree_depth"),
+      minLeafCount = cfg.getInt("min_leaf_count"))
+
+    val result: Array[((String, String), Any)] = getFeatures(
+      sc, input, minCount, sample, linearFeatureFamilies, options)
+    result
+  }
+
+  def getFeatures(sc: SparkContext, input: RDD[Example], minCount: Int, sample: Double,
+                  linearFeatureFamilies: util.List[String],
+                  options: NDTreeBuildOptions): Array[((String, String), Any)] = {
+    val linearFeatureFamiliesBC = sc.broadcast(linearFeatureFamilies)
+    val optionsBC = sc.broadcast(options)
+    val tree: Array[((String, String), Any)] = input.sample(true, sample).mapPartitions(partition => {
+      flattenExample(partition, linearFeatureFamiliesBC.value)
+    }).reduceByKey((a, b) => {
+      if (a.isInstanceOf[ArrayBuffer[Array[Double]]]) {
+        a.asInstanceOf[ArrayBuffer[Any]].++=(b.asInstanceOf[ArrayBuffer[Any]])
+      } else if (a.isInstanceOf[FeatureStats]) {
+        val as = a.asInstanceOf[FeatureStats]
+        val bs = b.asInstanceOf[FeatureStats]
+        FeatureStats(as.count + bs.count,
+          scala.math.min(as.min, bs.min),
+          scala.math.max(as.max, bs.max))
+      } else {
+        log.error(s"reduceByKey unknow ${a}")
+      }
+    }).filter(x => {
+      val a = x._2
+      if (a.isInstanceOf[ArrayBuffer[Array[Double]]]) {
+        a.asInstanceOf[ArrayBuffer[Any]].size >= minCount
+      } else if (a.isInstanceOf[FeatureStats]) {
+        a.asInstanceOf[FeatureStats].count >= minCount
+      } else {
+        log.error(s"filter unknown ${a}")
+        false
+      }
+    }).map(x => {
+      val a = x._2
+      if (a.isInstanceOf[ArrayBuffer[Array[Double]]]) {
+        // build tree
+        ((x._1._1, x._1._2), NDTree(optionsBC.value,
+          x._2.asInstanceOf[ArrayBuffer[Array[Double]]].toArray).model)
+      } else {
+        x
+      }
```

Comment:
I feel there are too many `asInstanceOf` and `Any` in the above code, there should be better way to write